samza-api/src/main/java/org/apache/samza/storage/StorageEngine.java

```diff
@@ -43,11 +43,15 @@
    * provided in one {@link java.util.Iterator} and not deserialized for
    * efficiency, allowing the implementation to optimize replay, if possible.
    *
+   * The implementers are expected to handle interrupt signals to the restoration thread and rethrow the exception to
```

Comment:
1. Yes, from my initial investigation, restore required handling. However, the pluggable nature of components can require interrupt handling in other components too. We will need to come up with enforcing this pattern at least to the newly added components and retroactively fix the old ones.
2. Good catch! Looks like checked exception addition does break source code/contractual compatibility even though it is binary compatible. Let me update the API notes in description.
3. Let me add additional documentation around handling of interrupts and provide some guidance.
4. Agreed. long term, it would be good to consolidate all of it and have some notion of interrupt aware operation/function and enforce it as programming paradigm for long running components or use shutdown aware patterns sorts.

Reply:
1-3. Sounds good.
4. I like a more explicit shutdown pattern. If that is determined to be out of scope for now, then could you please write down the idea in a JIRA so we can start tracking it?

Reply:
Yup. Will create a JIRA to capture the ideas.

Reply:
Added some more documentation and created the follow up ticket to capture shutdown aware idea - https://issues.apache.org/jira/browse/SAMZA-2414